### PR TITLE
[FrameworkBundle] Use relative paths in templates paths cache

### PR DESCRIPTION
**Pull request by tgalopin**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | yes |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #3079 |
| License | MIT |
| Doc PR | - |

This implements the usage of relative paths instead of absolute ones in `var/cache/*/templates.php`, important for ability to build the cache in a different context than where it will be used.

This PR transforms the following `templates.php`:

``` php
<?php return array (
  ':default:index.html.twig' => '/home/tgalopin/www/symfony-standard/app/Resources/views/default/index.html.twig',
  '::base.html.twig' => '/home/tgalopin/www/symfony-standard/app/Resources/views/base.html.twig',
);
```

Into:

``` php
<?php return array (
  ':default:index.html.twig' => __DIR__.'/../../../app/Resources/views/default/index.html.twig',
  '::base.html.twig' => __DIR__.'/../../../app/Resources/views/base.html.twig',
);
```

I also added tests for the TemplateCachePathsWarmer and improved tests for the TemplateLocator.

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/Templating/Loader/TemplateLocator.php [hunk 1]
```diff
@@ -23,20 +23,25 @@ class TemplateLocator implements FileLocatorInterface
 {
     protected $locator;
     protected $cache;
+    protected $rootDir;
```
Comment (hhamon) [inline src/Symfony/Bundle/FrameworkBundle/Templating/Loader/TemplateLocator.php]: should be private I think.

**Comment by tgalopin** [comment]
Travis failure is not related (PHP 7.1 segfault).

File: src/Symfony/Bundle/FrameworkBundle/Templating/Loader/TemplateLocator.php [hunk 2]
```diff
@@ -72,13 +77,29 @@ public function locate($template, $currentPath = null, $first = true)
...
-            return $this->cache[$key];
+            return '.' === $this->cache[$key][0] ? $this->rootDir.$this->cache[$key] : $this->cache[$key];
         }
 
         try {
-            return $this->cache[$key] = $this->locator->locate($template->getPath(), $currentPath);
+            $path = $this->locator->locate($template->getPath(), $currentPath);
+
+            if ($this->rootDir && 0 === strpos($path, $this->rootDir)) {
+                return $this->rootDir.($this->cache[$key] = './'.substr($path, strlen($this->rootDir)));
```
Comment (sstok) [inline src/Symfony/Bundle/FrameworkBundle/Templating/Loader/TemplateLocator.php]: Tip: As the length of the rootDir doesn't change at runtime you can cache it as a property or static local variable.

File: src/Symfony/Bundle/FrameworkBundle/Templating/Loader/TemplateLocator.php [hunk 3]
```diff
@@ -23,20 +23,25 @@ class TemplateLocator implements FileLocatorInterface
...
-     * @param FileLocatorInterface $locator  A FileLocatorInterface instance
-     * @param string               $cacheDir The cache path
+     * @param FileLocatorInterface $locator       A FileLocatorInterface instance
+     * @param string               $cacheDir      The cache path
+     * @param string               $kernelRootDir
      */
-    public function __construct(FileLocatorInterface $locator, $cacheDir = null)
+    public function __construct(FileLocatorInterface $locator, $cacheDir = null, $kernelRootDir = null)
     {
+        $this->cache = array();
```
Comment (sstok) [inline src/Symfony/Bundle/FrameworkBundle/Templating/Loader/TemplateLocator.php]: Better to set this when the property is defined.

**Comment by sstok** [comment]
LGTM.

**Comment by nicolas-grekas** [comment]
:+1:

File: src/Symfony/Bundle/FrameworkBundle/Templating/Loader/TemplateLocator.php [hunk 4]
```diff
@@ -32,7 +32,7 @@ class TemplateLocator implements FileLocatorInterface
      */
     public function __construct(FileLocatorInterface $locator, $cacheDir = null)
     {
-        if (null !== $cacheDir && is_file($cache = $cacheDir.'/templates.php')) {
+        if (null !== $cacheDir && file_exists($cache = $cacheDir.'/templates.php')) {
```
Comment (dunglas) [inline src/Symfony/Bundle/FrameworkBundle/Templating/Loader/TemplateLocator.php]: Why this change?

**Comment by dunglas** [comment]
👍

File: src/Symfony/Bundle/FrameworkBundle/Templating/Loader/TemplateLocator.php (hunk 4 above)
Comment (tgalopin) [inline src/Symfony/Bundle/FrameworkBundle/Templating/Loader/TemplateLocator.php]: file_exists is faster and is_file have no advantage over it here.

**Comment by fabpot** [comment]
Thank you @tgalopin.
